// cf-style-container/lib/applyStaticStyles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDisplayName = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactFela = require('react-fela');

var _displayName = require('./displayName');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable behance/no-deprecated */
const getDisplayName = exports.getDisplayName = (0, _displayName.hocDisplayName)('WithStaticStyles'); // eslint-disable-line behance/no-deprecated


const applyStaticStyles = (staticStyles, ComponentToWrap) => {
  class CompWithStaticStyles extends _react.PureComponent {
    componentWillMount() {
      const contextTheme = this.props.theme || {};
      staticStyles = typeof staticStyles === 'function' ? staticStyles(contextTheme) : staticStyles;
      this.context.renderer.renderStatic(staticStyles);
    }

    render() {
      return _react2.default.createElement(ComponentToWrap, this.props);
    }
  }

  CompWithStaticStyles.contextTypes = {
    renderer: _propTypes2.default.object
  };

  CompWithStaticStyles.propTypes = {
    theme: _propTypes2.default.object
  };

  const WithThemeHOC = (0, _reactFela.withTheme)(CompWithStaticStyles);
  (0, _displayName.addDisplayName)(WithThemeHOC, ComponentToWrap, getDisplayName);

  return WithThemeHOC;
};

exports.default = applyStaticStyles;
