// cf-style-container/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createStyledComponent = exports.getApplyStatisticStylesDisplayName = exports.applyStaticStyles = exports.getWithRendererDisplayName = exports.withRenderer = exports.withTheme = exports.capitalize = exports.getCreateComponentStylesDisplayName = exports.createComponentStyles = exports.combineRules = exports.connect = exports.ThemeProvider = exports.getApplyThemeDisplayName = exports.applyTheme = exports.getCreateComponentDisplayName = exports.createComponent = undefined;

var _fela = require('fela');

var _reactFela = require('react-fela');

var _underscore = require('underscore.string');

var _applyStaticStyles = require('./applyStaticStyles');

var _applyStaticStyles2 = _interopRequireDefault(_applyStaticStyles);

var _applyTheme = require('./applyTheme');

var _applyTheme2 = _interopRequireDefault(_applyTheme);

var _createComponent = require('./createComponent');

var _createComponent2 = _interopRequireDefault(_createComponent);

var _createComponentStyles = require('./createComponentStyles');

var _createComponentStyles2 = _interopRequireDefault(_createComponentStyles);

var _createStyledComponent = require('./createStyledComponent');

var _createStyledComponent2 = _interopRequireDefault(_createStyledComponent);

var _withRenderer = require('./withRenderer');

var _withRenderer2 = _interopRequireDefault(_withRenderer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-line behance/no-deprecated
exports.createComponent = _createComponent2.default;
exports.getCreateComponentDisplayName = _createComponent.getDisplayName;
exports.applyTheme = _applyTheme2.default;
exports.getApplyThemeDisplayName = _applyTheme.getDisplayName;
exports.ThemeProvider = _reactFela.ThemeProvider;
exports.connect = _reactFela.connect;
exports.combineRules = _fela.combineRules;
exports.createComponentStyles = _createComponentStyles2.default;
exports.getCreateComponentStylesDisplayName = _createComponentStyles.getDisplayName;
exports.capitalize = _underscore.capitalize;
exports.withTheme = _reactFela.withTheme;
exports.withRenderer = _withRenderer2.default;
exports.getWithRendererDisplayName = _withRenderer.getDisplayName;
exports.applyStaticStyles = _applyStaticStyles2.default;
exports.getApplyStatisticStylesDisplayName = _applyStaticStyles.getDisplayName;
exports.createStyledComponent = _createStyledComponent2.default; // eslint-disable-line behance/no-deprecated
